File: App/app/src/main/java/com/cinema/ticket_booking/data/repository/UserRepository.java
package com.cinema.ticket_booking.data.repository;

import androidx.lifecycle.*;
import com.cinema.ticket_booking.data.model.response.*;
import com.cinema.ticket_booking.network.ApiService;
import com.cinema.ticket_booking.util.Resource;
import retrofit2.Call;
import retrofit2.Callback;
import retrofit2.Response;
import javax.inject.Inject;
import javax.inject.Singleton;

@Singleton
public class UserRepository {
    private final ApiService api;

    @Inject
    public UserRepository(ApiService api) {
        this.api = api;
    }

    public LiveData<Resource<UserResponse>> getProfile() {
        MutableLiveData<Resource<UserResponse>> r = new MutableLiveData<>();
        r.setValue(Resource.loading());
        api.getMyProfile().enqueue(new Callback<>() {
            @Override
            public void onResponse(Call<ApiResponse<UserResponse>> c, Response<ApiResponse<UserResponse>> res) {
                if (res.isSuccessful() && res.body() != null && res.body().isSuccess())
                    r.setValue(Resource.success(res.body().getData()));
                else
                    r.setValue(Resource.error("Tải profile thất bại"));
            }

            @Override
            public void onFailure(Call<ApiResponse<UserResponse>> c, Throwable t) {
                r.setValue(Resource.error("Lỗi kết nối: " + t.getMessage()));
            }
        });
        return r;
    }

    public LiveData<Resource<UserResponse>> updateNotificationSettings(Boolean marketing, Boolean transaction) {
        MutableLiveData<Resource<UserResponse>> r = new MutableLiveData<>();
        r.setValue(Resource.loading());
        com.cinema.ticket_booking.data.model.request.NotificationSettingsRequest req = 
            new com.cinema.ticket_booking.data.model.request.NotificationSettingsRequest(marketing, transaction);
            
        api.updateNotificationSettings(req).enqueue(new Callback<>() {
            @Override
            public void onResponse(Call<ApiResponse<UserResponse>> c, Response<ApiResponse<UserResponse>> res) {
                if (res.isSuccessful() && res.body() != null && res.body().isSuccess())
                    r.setValue(Resource.success(res.body().getData()));
                else
                    r.setValue(Resource.error("Lỗi cập nhật cài đặt"));
            }

            @Override
            public void onFailure(Call<ApiResponse<UserResponse>> c, Throwable t) {
                r.setValue(Resource.error("Lỗi kết nối: " + t.getMessage()));
            }
        });
        return r;
    }
}

File: App/app/src/main/java/com/cinema/ticket_booking/data/repository/UserRepository.kt
package com.cinema.ticket_booking.data.repository

import androidx.lifecycle.LiveData
import androidx.lifecycle.MutableLiveData
import com.cinema.ticket_booking.data.model.request.NotificationSettingsRequest
import com.cinema.ticket_booking.data.model.response.*
import com.cinema.ticket_booking.network.ApiService
import com.cinema.ticket_booking.util.Resource
import javax.inject.Inject
import javax.inject.Singleton
import retrofit2.Call
import retrofit2.Callback
import retrofit2.Response

@Singleton
class UserRepository @Inject constructor(
    private val api: ApiService
) {

    fun getMyProfile(): LiveData<Resource<UserResponse>> {
        val r = MutableLiveData<Resource<UserResponse>>()
        r.value = Resource.loading()
        api.getMyProfile().enqueue(object : Callback<ApiResponse<UserResponse>> {
            override fun onResponse(c: Call<ApiResponse<UserResponse>>, res: Response<ApiResponse<UserResponse>>) {
                if (res.isSuccessful && res.body() != null && res.body()!!.success) {
                    r.value = Resource.success(res.body()!!.data)
                } else {
                    r.value = Resource.error("Tải profile thất bại")
                }
            }

            override fun onFailure(c: Call<ApiResponse<UserResponse>>, t: Throwable) {
                r.value = Resource.error("Lỗi kết nối: ${t.message}")
            }
        })
        return r
    }

    // Alias for backward compatibility
    fun getProfile(): LiveData<Resource<UserResponse>> {
        return getMyProfile()
    }

    fun updateNotificationSettings(marketing: Boolean?, transaction: Boolean?): LiveData<Resource<UserResponse>> {
        val r = MutableLiveData<Resource<UserResponse>>()
        r.value = Resource.loading()
        val req = NotificationSettingsRequest(marketing, transaction)

        api.updateNotificationSettings(req).enqueue(object : Callback<ApiResponse<UserResponse>> {
            override fun onResponse(c: Call<ApiResponse<UserResponse>>, res: Response<ApiResponse<UserResponse>>) {
                if (res.isSuccessful && res.body() != null && res.body()!!.success) {
                    r.value = Resource.success(res.body()!!.data)
                } else {
                    r.value = Resource.error("Lỗi cập nhật cài đặt")
                }
            }

            override fun onFailure(c: Call<ApiResponse<UserResponse>>, t: Throwable) {
                r.value = Resource.error("Lỗi kết nối: ${t.message}")
            }
        })
        return r
    }

    fun redeemGiftCard(code: String): LiveData<Resource<GiftCardResponse>> {
        val r = MutableLiveData<Resource<GiftCardResponse>>()
        r.value = Resource.loading()
        val body = HashMap<String, String>().apply {
            put("code", code)
        }
        api.redeemGiftCard(body).enqueue(object : Callback<ApiResponse<GiftCardResponse>> {
            override fun onResponse(c: Call<ApiResponse<GiftCardResponse>>, res: Response<ApiResponse<GiftCardResponse>>) {
                if (res.isSuccessful && res.body() != null && res.body()!!.success) {
                    r.value = Resource.success(res.body()!!.data)
                } else {
                    r.value = Resource.error(
                        if (res.body() != null) res.body()!!.message else "Mã thẻ không hợp lệ"
                    )
                }
            }

            override fun onFailure(c: Call<ApiResponse<GiftCardResponse>>, t: Throwable) {
                r.value = Resource.error("Lỗi kết nối: ${t.message}")
            }
        })
        return r
    }
}
